File: nvpl_blas/f77/dasum.f
!******************************************************************************
! Content:
!     This example demonstrates use of API as below:
!     DASUM 
!
!******************************************************************************
      program   DASUM_MAIN
*
      implicit none

      include "nvpl_blas_f77_blas.fi"

*       Intrinsic Functions
      intrinsic  abs

      integer          n, incx, i
      parameter        (n=5, incx=1)
      double precision result
      integer          xsize
      parameter        (xsize=1+(n-1)*abs(incx))
      double precision x(xsize)
      data             (x(i), i=1,xsize)/1.0,-3.7,4.3,0.0,5.0/

*       External Subroutines
      external         print_dvector 
*
*      Executable Statements
*
      print*
      print 99
*
*       Print input data
      print*
      print 100, n, incx
      call print_dvector(x,n,incx,'X')
*
*      Call DASUM subroutine
      result = DASUM(n,x,incx)
*
      print*
      print 101, result 

      stop
 99   format('Example: DASUM for computing the sum of the absolute'
     &       ' values of the elements of vector')
 100  format('#### args: n=',i1,', incx=',i1)
 101  format('The sum of the absolute values of the elements of'
     &       ' vector: ', f6.2)
      end

File: nvpl_blas/c/dasum.c
/******************************************************************************
 * Content: 
 *     This example demonstrates use of API as below:
 *     cblas_dasum
 *
 ******************************************************************************/
#include "example_helper.h"

int main() {
    double result;
    nvpl_int_t N = 5;
    double * X;
    nvpl_int_t incX = 1;

    printf("\nExample: cblas_dasum for computing the sum of the absolute values of the elements of vector\n\n");
    printf("#### args: n=%" PRId64 ", incx=%" PRId64 "\n", (int64_t)N, (int64_t)incX);

    nvpl_int_t len_x = 1 + (N - 1) * labs(incX);
    // allocate memory
    X = (double *)malloc(len_x * sizeof(double));

    // fill data
    fill_dvector(X, N, incX);

    // print input data
    print_dvector(X, N, incX, "X");

    // call cblas_dasum
    result = cblas_dasum(N, X, incX);

    // print result
    printf("\nThe sum of the absolute values of the elements of vector: %f\n", result);

    // release memory
    free(X);
    return 0;
}
